### PR TITLE
Entry.sh file for username, password authentication in docker (second phase)

### PR DESCRIPTION
**Pull request by fgalan**
This is the second phase of the merge for PR https://github.com/telefonicaid/iotagent-json/pull/380. The diff +127 -5 is a good checksum ensuring that both PRs has the same content.

E2E CI testing will be done on this PR in order to guarantee nothing is broken. If tests are ok, then it will be merged.

CC: @Jagatjot @jason-fox @dcalvoalonso @AlvaroVega

### CODE REVIEW

**Comment by fgalan** [comment]
I have got the following issue attempting to build at dockehub:

```
chmod: cannot access 'docker/entrypoint.sh': No such file or directory
Removing intermediate container 9f55ad042121
```

@Jagatjot could you have a look, please? If modifications are needed, please do them as PRs on this PR branch (i.e. using feature/366_pre_merge as base branch).

In addition, please ensure that the docker build process works with the Dockerfile (i.e. build it locally so we can detect additional problems that we would got on dockerhub).

**Comment by jason-fox** [comment]
@fgalan - how are you building from the branch? By default the `Dockerfile` will download from `master`. The build will fail since `docker/entrypoint.sh` can't be found there yet.

The solution is to tag the hash you want to test and build using the command line syntax found in the [Ultralight Twin](https://github.com/telefonicaid/iotagent-ul/pull/339#issuecomment-461902493)

> I've labelled the tip of the branch `TEST`. To test the docker container run
> 
> ```
> docker build -t iot-agent . --build-arg DOWNLOAD=TEST --no-cache
> ```

**Comment by fgalan** [comment]
In fact, I think this is related with the issue I created some time ago https://github.com/telefonicaid/iotagent-ul/issues/317. I forgot about it ;)

The problem is that the E2E CI system takes images from dockerhub. So, at the end, I need the image built on dockerhub. Of course I can build locally and push to dockerhub but I'd prefer if possible some sort of solution fully contained in dockerhub.

In last comment in the cited issue you said "It looks like there could be a fix soon on the Docker Hub side". Let's check if that "soon" (around 3 months ago) has become reality :)

**Comment by jason-fox** [comment]
> In last comment in the cited issue you said "It looks like there could be a fix soon on the Docker Hub side". Let's check if that "soon" (around 3 months ago) has become reality :)

Docker Hub and Docker Cloud are now integrated, so it should just be a matter of adding an appropriate build hook and replacing `master` with `$SOURCE_BRANCH`

* https://www.infoq.com/news/2019/01/new-docker-hub
* https://docs.docker.com/docker-hub/builds/advanced/#environment-variables-for-building-and-testing
* https://devopsunleashed.com/2018/03/24/docker-hub-build-hooks/
* https://github.com/telefonicaid/fiware-sinfonier/commit/b7ed5fa990cd7

It would make sense to add things like the githash or build version as `ENV` variables as well I guess.

**Comment by Jagatjot** [comment]
> I have got the following issue attempting to build this branch at dockerhub:
> 
> ```
> chmod: cannot access 'docker/entrypoint.sh': No such file or directory
> Removing intermediate container 9f55ad042121
> ```
> @Jagatjot could you have a look, please? If modifications are needed, please do them as PRs on this PR branch (i.e. using feature/366_pre_merge as base branch).
> 
> In addition, please ensure that the docker build process works with the Dockerfile (i.e. build it locally so we can detect additional problems that we would got on dockerhub).

@fgalan it is because at dockerhub it is searching for docker/entrypoint.sh at master branch and it is not present at master therefore it says : No such file or directory

**Comment by jason-fox** [comment]
@Jagatjot - Try porting over the webhook from the Ultralight twin:

https://github.com/jason-fox/iotagent-ul/commit/ad474d60

File: docker/Dockerfile [hunk 1]
```diff
@@ -3,6 +3,7 @@ FROM node:${NODE_VERSION}
 ARG GITHUB_ACCOUNT=telefonicaid
 ARG GITHUB_REPOSITORY=iotagent-json
 ARG DOWNLOAD=latest
+ARG SOURCE_BRANCH=master
```
Comment (fgalan) [inline docker/Dockerfile]: Should the usage of this new ARG be taken into accoun in section "How to build your own image" of the README.md file? What do you think @jason-fox ?

**Comment by fgalan** [comment]
@jason-fox I have ported the commits you mention and not it seems it is working. Thanks!

imagen

So, e2e regression can be done now. Keep tuned :)

**Comment by jason-fox** [comment]
> I have ported the commits you mention and not it seems it is working. Thanks!
>
> So, e2e regression can be done now. Keep tuned :)

*not* or *now* ? I hope it is the latter.

**Comment by jason-fox** [comment]
Related to: https://github.com/telefonicaid/iotagent-node-lib/pull/767 - with the alternative **Keyrock** authentication, `IOTA_AUTH_CLIENT_ID` and `IOTA_AUTH_CLIENT_SECRET need to be added to the list:

line 46 of [entrypoint.sh](https://github.com/telefonicaid/iotagent-json/pull/390/files#diff-1fd0da033753906a29729c3b71ee9e6f) needs the following:

```bash
file_env 'IOTA_AUTH_USER'
file_env 'IOTA_AUTH_PASSWORD'
file_env 'IOTA_AUTH_CLIENT_ID'
file_env 'IOTA_AUTH_CLIENT_SECRET'
```

The commented out config section should be amended as well.

**Comment by fgalan** [comment]
> **not** or **now** ? I hope it is the latter.

Yes. I was now ;) In fact, the first round seems to be very promissing, but a new round will take place after the changes you have suggested at https://github.com/telefonicaid/iotagent-json/pull/390#issuecomment-477963004

On the meanwhile, please have a look to [this comment](https://github.com/telefonicaid/iotagent-json/pull/390#pullrequestreview-220151704).

**Comment by fgalan** [comment]
>line 46 of [entrypoint.sh](https://github.com/telefonicaid/iotagent-json/pull/390/files#diff-1fd0da033753906a29729c3b71ee9e6f) needs the following:
>
>```bash
>file_env 'IOTA_AUTH_USER'
>file_env 'IOTA_AUTH_PASSWORD'
>file_env 'IOTA_AUTH_CLIENT_ID'
>file_env 'IOTA_AUTH_CLIENT_SECRET'
>```

The four added in https://github.com/telefonicaid/iotagent-json/pull/390/commits/a35b1631cbbf7c96bf6e071acbc88ab29139a716 (then fixed in https://github.com/telefonicaid/iotagent-json/pull/390/commits/9dbff11778e4a973e4acca8656664830530c5853, as some of then already were there).

> The commented out config section should be amended as well.

Add (the two actually new) in https://github.com/telefonicaid/iotagent-json/pull/390/commits/f632cab43410cc84f9c68ef547900609f795dd30
Added in

File: docker/Dockerfile (hunk 1 above)
Comment (jason-fox) [reply docker/Dockerfile]: Yes - it should now be split into three parts.

see - [Ultralight](https://github.com/jason-fox/iotagent-ul/commit/ac295c05c9bffff0e3611eb85e4561b1c25243fb#diff-4b5e6a0d7df359d2d18257ca84179db7fd61ef67)


## How to build an image

--- etc

## Building from your own fork

To download code from your own fork of the GitHub repository add the  arguments to the `docker build` command.	

```console
docker build -t iot-agent . \
    --build-arg GITHUB_ACCOUNT=<your account> \
    --build-arg GITHUB_REPOSITORY=<your repo> \
    --build-arg SOURCE_BRANCH=<your branch>
```

 ## Building from your own source files

To download code from your own fork of the GitHub repository add the `GITHUB_ACCOUNT`, `GITHUB_REPOSITORY` and `SOURCE_BRANCH` arguments (default `master`) to the `docker build` command.

Comment (fgalan) [reply docker/Dockerfile]: Fixed in https://github.com/telefonicaid/iotagent-json/pull/390/commits/aef402558bf56ed213c85cee45b6be416f8efdea (by imitation of https://github.com/jason-fox/iotagent-ul/commit/ac295c05c9bffff0e3611eb85e4561b1c25243fb#diff-4b5e6a0d7df359d2d18257ca84179db7fd61ef67)

Comment (jason-fox) [reply docker/Dockerfile]: 👍 LGTM

**Pull request merged**